### PR TITLE
plot/vg/draw: introduce and use TextHandler

### PR DESCRIPTION
**Pull request by sbinet**
Please take a look.

### CODE REVIEW

**Comment by sbinet** [comment]
ping?

**Comment by kortschak** [comment]
This will take some time to look through. I will try to get to it in the next couple of days (task added to personal assistant).

**Comment by sbinet** [comment]
I've dropped the HTML handler out of this PR (to reduce the number of files to review)
If we proceed with this PR (or some version of it), I'll file it back in another one.

(I've also rebased off the latest master, and removed the `panic`s for the `Canvas.DPI` implementations (putting a default/hardcoded 72 value) for canvas implementations that do not expose/use an explicit DPI value)

File: gob/gob.go [hunk 1]
```diff
@@ -43,4 +44,7 @@ func init() {
 	gob.Register(plotter.XYZs{})
 	gob.Register(plotter.XYValues{})
 
+	// vg/draw.TextStyle
+	gob.Register(plot.DefaultTextHandler)
+	gob.Register(text.Plain{})
```
Comment (kortschak) [inline gob/gob.go:49]: `plot.DefaultTextHandler == text.Plain{}`; why do we register both?

File: text/latex.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/go-latex/latex/drawtex"
+	"github.com/go-latex/latex/font/ttf"
+	"github.com/go-latex/latex/mtex"
+	"github.com/go-latex/latex/tex"
+	"gonum.org/v1/plot/vg"
```
Comment (kortschak) [inline text/latex.go:15]: Insert blank line.

File: text/latex.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	sty draw.TextStyle
+	pt  vg.Point
+}
+
+func (r *latex) Render(width, height, dpi float64, c *drawtex.Canvas) error {
+	//var (
+	//	w = width * dpi
+	//	h = height * dpi
+	//)
+
```
Comment (kortschak) [inline text/latex.go:91]: Delete?

File: text/latex.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		{X: x1, Y: y1},
+	}
+
+	r.cnv.FillPolygon(r.sty.Color, pts)
+}
+
+var (
+	_ draw.TextHandler = (*Latex)(nil)
+	_ mtex.Renderer    = (*latex)(nil)
+)
```
Comment (kortschak) [inline text/latex.go:143]: Can you move this to the top? It should be a documentation of the interface that the type is intended to satisfy.

File: text/plain.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package text // import "gonum.org/v1/plot/text"
+
+import (
+	"gonum.org/v1/plot/vg/draw"
+)
```
Comment (kortschak) [inline text/plain.go:9]: `import "gonum.org/v1/plot/vg/draw"`

File: text/plain.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+// Plain is a text/plain handler.
+type Plain = draw.PlainTextHandler
+
+var (
+	_ draw.TextHandler = (*Plain)(nil)
+)
```
Comment (kortschak) [inline text/plain.go:16]: `var _ draw.TextHandler = (*Plain)(nil)`

File: text/text.go [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
```
Comment (kortschak) [inline text/text.go:1]: I think usually we call these "doc.go".

File: vg/recorder/recorder.go [hunk 8]
```diff
@@ -112,6 +112,11 @@ func (c *Canvas) append(a Action) {
 	c.Actions = append(c.Actions, a)
 }
 
+func (c *Canvas) DPI() float64 {
+	// FIXME(sbinet): add a way to specify a value?
```
Comment (kortschak) [inline vg/recorder/recorder.go:116]: If we need to add these we can either make canvases require a constructor which takes a DPI, or add a resolution field that is interpreted as 72 if it's 0, or make it a conditional interface where 72 is the default value if there is no DPI method. Given that LaTeX is the only one that needs it, the last approach feels the nicest to me.

File: vg/vg.go [hunk 9]
```diff
@@ -17,6 +17,8 @@ import (
 // graphics.
 // The origin is in the bottom left corner.
 type Canvas interface {
+	DPI() float64
+
```
Comment (kortschak) [inline vg/vg.go:21]: Why did we need to add the `DPI` method?

File: gob/gob.go (hunk 1 above)
Comment (sbinet) [reply gob/gob.go:49]: mainly for "forward coding" (in the perhaps remote case when we'd switch DefaultTextHandler to default to another implementation)

I can remove if you prefer.

File: vg/vg.go (hunk 9 above)
Comment (sbinet) [reply vg/vg.go:21]: for the text.Latex handler.

File: vg/recorder/recorder.go (hunk 8 above)
Comment (sbinet) [reply vg/recorder/recorder.go:116]: I am a bit worried a `xyz.Canvas` implementing this `DPI()` method could be wrapped by a couple layers of `vg.Canvas` (sans the `DPI()` method) interfaces and then presented to a `draw.Canvas` value.
I think the type switch wouldn't be able to dig that `DPI()` method out of the interface layers.

but perhaps it's just a remote, theoretical and uncommon failure mode....

File: gob/gob.go (hunk 1 above)
Comment (kortschak) [reply gob/gob.go:49]: OK. That's fine.

**Comment by sbinet** [comment]
ping?

**Comment by kortschak** [comment]
I'm still concerned about the `DPI` method.

**Comment by sbinet** [comment]
PTAL: removed DPI from vg.Canvas, set as a field of Latex.

File: text/latex.go [hunk 10]
```diff
@@ -18,15 +18,17 @@ import (
 )
 
 // Latex parses, formats and renders LaTeX.
-type Latex struct{}
+type Latex struct {
+	DPI float64
```
Comment (kortschak) [inline text/latex.go:22]: Add a doc comment that this sets that LaTeX font resolution and also that resolution defaults to 72 if this is unset.

Comment (sbinet) [reply text/latex.go:25]: done

**Pull request merged**